src/metabase/db.clj

```diff
@@ -43,25 +41,23 @@
   db-type
   quoting-style])
 
-(defonce ^:private db-setup-finished?
-  (atom false))
-
+;; TODO -- consider whether we can just do this automatically when `getConnection` is called on
+;; [[mdb.connection/*application-db*]] (or its data source)
 (defn db-is-set-up?
   "True if the Metabase DB is setup and ready."
-  ^Boolean []
-  @db-setup-finished?)
+  []
+  (= @(:status mdb.connection/*application-db*) ::setup-finished))
 
 (defn setup-db!
   "Do general preparation of database by validating that we can connect. Caller can specify if we should run any pending
   database migrations. If DB is already set up, this function will no-op. Thread-safe."
   []
-  (when-not @db-setup-finished?
-    (locking db-setup-finished?
-      (when-not @db-setup-finished?
+  (when-not (db-is-set-up?)
+    (locking (:status mdb.connection/*application-db*)
```

Comment:
this locking looks super weird to me.

Reply:
Yeah I just picked that arbitrarily to be the lock, I wanted to pick something related to `*application-db*` so that one application DB wouldn't be locked waiting for another one to finish setup. I guess maybe making `*application-db*` itself the lock would be better

Reply:
fixed